src/wagtail_vector_index/consumers.py

```diff
@@ -0,0 +1,149 @@
+import asyncio
+import logging
+
+from django.apps import apps
+from django.core.exceptions import PermissionDenied
+from channels.generic.http import AsyncHttpConsumer
+
+logger = logging.Logger(__name__)
+
+
+class WagtailVectorIndexSSEConsumer(AsyncHttpConsumer):
+    """
+    A Django Channels consumer for handling Server-Sent Events (SSE) related to WagtailVectorIndex queries.
+
+    Attributes:
+        page_instance (Model): The Wagtail page model instance for which the vector index is queried.
+        page_model_name (str): The name of the Wagtail page model for which the vector index is queried.
+        vector_index (VectorIndex): The vector index associated with the Wagtail page model.
+
+    Methods:
+        handle: The main entry point for processing HTTP requests, including SSE connections.
+        process_prompt: Processes the incoming prompt and sends SSE updates.
+        check_permissions: Checks user authentication and raises PermissionDenied if not authenticated.
+        ratelimit_request: Placeholder for implementing rate-limiting logic.
+
+    Usage:
+    - Inherit from this class and set either 'page_instance' or 'page_model_name' attribute.
+    - Implement custom logic within 'process_prompt' to handle vector index queries and SSE updates.
+    - Optionally, override 'check_permissions' and 'ratelimit_request' for additional safety checks.
+
+    Example:
+    ```python
+    class CustomSSEConsumer(WagtailVectorIndexSSEConsumer):
+        page_instance = YourWagtailPageModel
+
+        async def process_prompt(self, query):
+            # Your custom logic to handle the query and send SSE updates
+            pass
+    ```
+    """
+    page_instance = None
+    page_model_name = None
+
+    def __init__(self, *args, **kwargs):
```

Comment:
**thought:** I think this would be a lot more flexible if it didn't need subclassing for each index, and didn't need to be aware of Pages/models.

If the user could pass a vector index class name along with their request, you could look it up in the [registry](https://github.com/wagtail/wagtail-vector-index/blob/main/src/wagtail_vector_index/index/registry.py) and use it directly from there?

Reply:
Updated, although uneasy with how I've updated the code to use `apps.get_model()` it doesn't seem like a secure way of doing it but unsure of alternatives.

Reply:
Closing this in favor of https://github.com/wagtail/wagtail-vector-index/pull/28#discussion_r1434249106